File: libraries/hlp/hdread.c
#include "hlpsys.h"
int hlpHdread ( int lstring, long *iadr, char *string, int *nc )
/*
**  - - - - - - - - - -
**   h l p H d r e a d
**  - - - - - - - - - -
**
**  Direct-access read from the HELP library.
**
**  Given (in global data):
**     jhelp     int        state of HELP system: 1=open/write
**     *fphl     FILE       file pointer for HELP library file
**     nchh      long       number of characters in HELP library file
**
**  Given (argument):
**     lstring   int        length of string
**     *iadr     long       character address within HELP file (1st = 0)
**
**  Returned (argument)
**     *iadr     long       points to next character in sequence
**     *string   char       input record
**     *nc       int        length of record (0 or more)
**
**  Returned (function value):
**               int        status:  0 = OK
**                   hlp_ILLEGAL_STATE = HELP system in wrong state
**                      hlp_READ_ERROR = read error
**                       hlp_READ_WIDE = attempt to read outside file
**                 hlp_RECORD_OVERSIZE = record overflows STRING (note 2)
**
**  The above error codes are defined in #include file hlpsys.h.
**
**  Notes:
**
**  1)  No check is made that *iadr points to the start of a record.
**
**  2)  If the record overflows string, the first lstring-1 characters
**      are stored in string followed by '\0', nc is set to lstring-1
**      and an error status is returned.
**
**  Last revision:   13 September 1995
**
**  Copyright 1996 P.T.Wallace.  All rights reserved.
*/

#define EOS '\0'

{
   int j, c;


/* Abort if the HELP system isn't in the right state. */
   if ( jhelp != 2 ) return hlp_ILLEGAL_STATE;

/* Abort if the output string isn't big enough at least for the EOS. */
   if ( lstring < 1 ) return hlp_RECORD_OVERSIZE;

/* Abort if the specified address lies outside the file. */
   if ( *iadr < 0 || *iadr >= nchh ) return hlp_READ_WIDE;

/* Locate the right place in the file. */
   if ( fseek ( fphl, *iadr, 0 ) ) return hlp_READ_ERROR;

/* Initialize the character count (which is also the string index). */
   *nc = 0;

/* Preset status to OK. */
   j = 0;

/* Read characters until end-of-string or an error condition. */
   c = 'A';
   while ( c != EOS ) {

   /* Read the next character. */
      if ( ( c = fgetc ( fphl ) ) == EOF ) {

      /* Error: set status and prepare to terminate string. */
         j = hlp_READ_ERROR;
         c = EOS;
      } else {

      /* Store the character and increment the string index. */
         string [ ( *nc ) ++ ] = (char) c;

      /* Was that the EOS? */
         if ( c != EOS ) {

         /* No: is string prematurely full? */
            if ( *nc >= lstring ) {

            /* Yes: replace the last character with EOS and set status. */
               c = EOS;
               string [ *nc - 1 ] = (char) c;
               j = hlp_RECORD_OVERSIZE;

          /* Or are we about to drop off the end of the file? */
            } else if ( ( *iadr ) ++ >= nchh ) {

            /* Yes: replace the last character with EOS and set status. */
               c = EOS;
               string [ *nc - 1 ] = (char) c;
               j = hlp_READ_WIDE;
            }
         }
      }
   }

/* Adjust the character count to exclude the EOS character. */
   ( *nc ) --;

/* Push the address to the start of the next record. */
   ( *iadr ) ++;

/* Return the final status. */
   return j;
}

File: libraries/hlp/hdread.f
      SUBROUTINE hlp_HDREAD (IADR, STRING, NC, J)
*+
*  - - - - - - -
*   H D R E A D
*  - - - - - - -
*
*  Direct-access read from the help library.
*
*  Specified in INCLUDE:
*     LHBUF     i          length of help buffer
*
*  Given (in COMMON):
*     JHELP     i          state of HLP system: 1=open/write
*     LUHL      i          unit number for help library file
*     IHBUF     i          address of HBUF within help library file
*     NCHH      i          number of characters in help library file
*     HBUF      c*(LHBUF)  help buffer
*     HEOS      c*1        end-of-string character
*
*  Given (argument):
*     IADR      i          character address within help file (1st = 0)
*
*  Returned (in COMMON):
*     HBUF      c*(LHBUF)  help buffer
*     IHBUF     i          address of HBUF within help file
*
*  Returned (argument)
*     IADR      i          points to next character in sequence
*     STRING    c*(*)      input record (not including end-of-string)
*     NC        i          length of record (0 or more)
*     J         i          status:  0 = OK
*                                  -1 = HLP system in wrong state
*                                  -4 = read error
*                                  -7 = attempt to read outside file
*                                  -8 = record overflows STRING (note 2)
*
*  Notes:
*
*  1)  No check is made that IADR points to the start of a record.
*
*  2)  If the record overflows STRING, the first LEN(STRING) characters
*      are stored in STRING, NC is set to LEN(STRING) and J is set to
*      -8.
*
*  P.T.Wallace   Starlink   13 June 1995
*-

      IMPLICIT NONE

      INTEGER IADR
      CHARACTER*(*) STRING
      INTEGER NC,J

      INCLUDE 'helpic'

*  Length of STRING
      INTEGER L

*  Character index to STRING
      INTEGER IS

*  Loop control
      LOGICAL MORE

*  Character address within help library file
      INTEGER IC

*  Direct-access record number
      INTEGER IREC

*  Character index to input buffer BUF
      INTEGER I

*  Single character
      CHARACTER C



*  Check that the HLP system is in the right state.
      IF (JHELP.NE.2) GO TO 9000

*  Preset status to OK.
      J=0

*  Length of STRING.
      L=LEN(STRING)

*  Fill buffer if uninitialized or past the specifed place.
      IF (IHBUF.LT.0.OR.IADR.LT.IHBUF) THEN
         IREC=1+IADR/LHBUF
         READ (UNIT=LUHL,FMT='(A)',REC=IREC,ERR=9020) HBUF

*     Character address of start of buffer.
         IHBUF=LHBUF*(IREC-1)
      END IF

*  Initialize STRING pointer.
      IS=1

*  Input characters until EOS or STRING full.
      MORE=.TRUE.
      DO WHILE (MORE)

*     Current character address.
         IC=IADR+IS-1

*     Error if beyond end of help library file.
         IF (IC.GE.NCHH) GO TO 9010

*     Is the current character within the current buffer?
         IF (IC.GE.IHBUF+LHBUF) THEN

*        No: read the next record.
            IREC=1+IC/LHBUF
            READ (UNIT=LUHL,FMT='(A)',REC=IREC,ERR=9020) HBUF

*        Character address of start of buffer.
            IHBUF=LHBUF*(IREC-1)
         END IF

*     We have the correct record in the buffer: extract character.
         I=IC-IHBUF+1
         C=HBUF(I:I)

*     EOS?
         IF (C.EQ.HEOS) THEN

*        Yes: done.
            MORE=.FALSE.
         ELSE

*        No: is STRING already full?
            IF (IS.LE.L) THEN

*           No: copy the character into STRING.
               STRING(IS:IS)=C
               IS=IS+1
            ELSE

*           STRING full: set error status.
               J=-8
            END IF
         END IF

*     Next character.
      END DO

*  Spacefill STRING.
      IF (IS.LE.L) STRING(IS:)=' '

*  Update IADR to point to next record.
      IADR=IC+1

*  Number of characters transferred.
      NC=IS-1

*  Exit.
      GO TO 9999

*  HLP system in wrong state
 9000 CONTINUE
      J=-1
      GO TO 9999

*  Attempt to read outside the file
 9010 CONTINUE
      J=-7
      GO TO 9999

*  Read error
 9020 CONTINUE
      J=-4

*  Exit.
 9999 CONTINUE

      END
